src/Autofac/Core/Resolving/ResolveOperationBase.cs

```diff
@@ -0,0 +1,263 @@
+
+            return result;
+        }
+
+        /// <inheritdoc />
+        public object GetOrCreateInstance(ISharingLifetimeScope currentOperationScope, ResolveRequest request)
+        {
+            if (_ended) throw new ObjectDisposedException(ResolveOperationResources.TemporaryContextDisposed, innerException: null);
+
+            // Create a new request context.
+            var requestContext = new ResolveRequestContext(this, request, currentOperationScope, _pipelineTracer);
+
+            // Raise our request-beginning event.
+            var handler = ResolveRequestBeginning;
+            handler?.Invoke(this, new ResolveRequestBeginningEventArgs(requestContext));
+
+            RequestDepth++;
+
+            // Track the last active request and scope in the call stack.
+            var lastActiveRequest = ActiveRequestContext;
+            var lastScope = CurrentScope;
+
+            ActiveRequestContext = requestContext;
+            CurrentScope = currentOperationScope;
+
+            try
+            {
+                _pipelineTracer?.RequestStart(this, requestContext);
+
+                ExecuteRequest(requestContext);
+
+                if (requestContext.Instance == null)
+                {
+                    // No exception, but was null; this shouldn't happen.
+                    throw new DependencyResolutionException(ResolveOperationResources.PipelineCompletedWithNoInstance);
+                }
+
+                _successfulRequests.Add(requestContext);
+                _pipelineTracer?.RequestSuccess(this, requestContext);
+            }
+            catch (Exception ex)
+            {
+                _pipelineTracer?.RequestFailure(this, requestContext, ex);
+                throw;
+            }
+            finally
+            {
+                ActiveRequestContext = lastActiveRequest;
+                CurrentScope = lastScope;
+
+                // Raise the appropriate completion events.
+                if (RequestStack.Count == 0)
+                {
+                    CompleteRequests();
+                }
+
+                RequestDepth--;
+            }
+
+            return requestContext.Instance;
+        }
+
+        /// <summary>
+        /// An implementation must implement this member to execute the specified request context.
+        /// </summary>
+        /// <param name="requestContext">The request context.</param>
+        protected abstract void ExecuteRequest(ResolveRequestContextBase requestContext);
+
+        private void CompleteRequests()
+        {
+            var completed = _successfulRequests;
+            int count = completed.Count;
+            ResetSuccessfulRequests();
+
+            for (int i = 0; i < count; i++)
+            {
+                completed[i].Complete();
+            }
+        }
+
+        private void ResetSuccessfulRequests()
+        {
+            _successfulRequests = new List<ResolveRequestContext>();
```

Comment:
I did have a look at swapping it for a `Clear`, left it as `new` for functionality reasons. If you look at the `CompleteRequests` method, it takes a reference to the original success list, then replaces it, and executes Complete on each original successful request.
Additional resolves can happen inside there, adding more successful requests to complete, which then has to almost go round in a loop.

Initialising to a pre-defined capacity is an option, but what would it be? The default capacity is 4 items.

Perhaps instead of resetting the list or clearing it, we could track the start and end position in the list that is consumed by the `CompleteRequests` method; subsequent invokes of `CompleteRequests` would then start from a later point in the list, ignoring the earlier successes.

Reply:
Hmmm. I see what you're saying. new may actually be the way to go. Let's leave it for now.